proposed/2023/nuget-package-search.md

```diff
@@ -0,0 +1,74 @@
+## Explanation
+
+### Functional explanation
+
+<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
+<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->
+#### 1. Adding `-ExactMatch` Option: 
+A new flag, `-ExactMatch`, will be added to the `search` command. When this flag is used, it will only return packages whose names exactly match the query.
+- **Version not specified** : In this case, the latest version of the package specified is listed .\
+*Nuget.exe Usage* : `nuget search <PackageName> -ExactMatch` \
+*Dotnet Usage* : `dotnet nuget search <PackageName> --exact-match`
+- **Version specified** : In this case, the package with the specific version number is listed.\
+*Nuget.exe Usage* : `nuget search <PackageName> -ExactMatch -Version 1.2.3.4` \
+*Dotnet Usage* : `dotnet nuget search <PackageName> --exact-match --version 1.2.3.4`
+- **All Versions** : In this case, all versions of the package with the exact match are listed.\
+*Nuget.exe Usage* : `nuget search <PackageName> -ExactMatch -AllVersions` \
+*Dotnet Usage* : `dotnet nuget search <PackageName> --exact-match --all-versions`
+
+If no exact match is found, display an error: `Error: An exact match for package <PackageName> is not found.`
+#### 2. Deprecating list Command: 
+The list command is planned to be deprecated, given its functionalities can be covered by the search command. A warning message will be displayed to inform users of this change. `Warning: The 'list' command is deprecated. Use 'search' instead.` \
+The functionalities of the `list` command can be replicated as follows
+* In nuget.exe `nuget list` will be `nuget search`.
+* In Dotnet, `dotnet nuget search` will be added and it will be the same as `nuget search`.
+#### 3. Introducing nuget search to Dotnet:
+A new command `dotnet nuget search` will be introduced to bring package search functionality to the Dotnet CLI. This will help in resolving the highly requested `dotnet nuget list` functionality.
+### Technical explanation
+
+<!-- Explain the proposal in sufficient detail with implementation details, interaction models, and clarification of corner cases. -->
+The `-ExactMatch` flag will require modifying the query parser for the `search` command in NuGet.
```

Comment:
Which HTTP resource will `-ExactMatch` use? For the [search resource](https://learn.microsoft.com/en-us/nuget/api/search-query-service-resource), nuget.org supports `packageid:My.Exact.Id` to return a single result in the search results. But this is not guaranteed to work across various package feeds. You can perhaps consider using the [package metadata](https://learn.microsoft.com/en-us/nuget/api/registration-base-url-resource) (aka registration) resource, you can get exact ID metadata but there is no guarantee the metadata available on search vs. package metadata is going to be the same.

For the client-side filtering option, the unfortunate reality is that a poorly implemented, although spec compliant search resource may not return an exact match in the first page of results. In fact, there were periods of time when exact match didn't come first (or on the first page of results) on NuGet.org! So I don't see client side filtering as ideal either.

Generally, I'm curious about the specific details of "require modifying the query parser".

Reply:
Do we have a pragmatic solution that works in every case? 

I'm wondering if we need to commit to a trade-off that's likely to cause the least amount of problems. 
IMO, using metadata resource and ignoring inconsistencies is likely to work the best. 

I think we should choose 1 and document others options though! Great call out!

Reply:
> IMO, using metadata resource and ignoring inconsistencies is likely to work the best.

Yes this is the way. You already have a unified model so strategically selecting output/display properties based on the intersection between search and registration seems like the right move.